ipwatchd/ipwatchd.8
.\" (c) 2006, Quest Software, Inc. All rights reserved.
.TH IPWATCHD 8
.SH NAME
ipwatchd \- runs a program when the primary IP address changes
.SH SYNOPSIS
.B ipwatchd
[\-d]
.RI [\-o name = value ]
.RI [ change_program ]
.SH DESCRIPTION
The
.B ipwatchd
daemon watches MacOS X's dynamic System Configuration state for interface
address changes, and runs the given program with the new IP address.
.PP
This program is intended to be run and managed by the
launchd(8)
daemon.
.PP
The
.B ipwatch
daemon watches the dynamic store string at
.I State:/Network/Global/IPv4[PrimaryService]
to determine what the current "primary" interface number is
(For example, "0", which is usually en0).
From this it watches the primary service's IP address(es) at
.I State:/Network/Service/$PrimaryService/IPv4[Addresses]
and uses the first IP address as the primary address of the host.
.PP
When the primary IP address change is detected,
.B ipwatchd
then waits for a period of
quiet (the "change delay") before running the
.I change_program
with the new IP address as its first and only argument.
.SH OPTIONS
.TP
.RI \-d
This option
enables debug (verbose) mode.
It is equivalent to
.B \-oDebug=true
.TP
.RI \-o name = value
Sets a configuration option by name as described in the section below.
.TP
.I change_program
This argument is the same as 
.BI \-oChangeProgram= change_program
.SS "Configuration Options"
The \-o option can be used to set named options.
Unknown options are ignored.
Known options are:
.TP
.RI Debug \ (boolean)
Enables debug messages sent to standard error.
.TP
.RI ChangeProgram \ (string)
The program to execute when the IP address changes. 
The executed program is given the new IP address as its first argument.
This defaults to /opt/quest/sbin/dnsupdate
.TP
.RI GlobalIPv4Key \ (string)
The name of the IPv4 key in the dynamic store.
Defaults to "State:/Network/Global/IPv4"
.TP
.RI PrimaryIPv4KeyPattern \ (string)
The pattern to use when detecting interfaces.
The "%" character in this string is replaced with the primary service ID.
Defaults to "State:/Network/Service/%/IPv4".
.TP
.RI ChangeDelay \ (float)
The delay in seconds to wait for quiescence after detecting an IP address 
and before running the ChangeProgram.
Values of zero or lower are interpreted as meaning immediate application.
Defaults to 30 seconds.
.SH FILES
.I /Library/LaunchDaemons/com.quest.rc.ipwatchd.plist
.SH "SEE ALSO"
dnsupdate(8),
launchd(8),
.I http://rc.vintela.com/topics/ddns/
.SH AUTHORS
Quest Software, Inc.


CHANGES
1.3.3 [2009-04-21]
    - Work around KRB5_FCC_INTERNAL error on Mac OS 10.5 when run from
      launchd (RC bug #663).
    - Don't use the "-a" auth domain for PTR updates (VAS bug #15832)
    - Portability bug fixes for HP-UX and Solaris
      (VAS bugs #15690, #15811, #15812, #15881)
    - Fixed uninstall problem on Mac OS X (VAS bug #15813)
    - Fix handling if gethostname() fails
    - Clean up temporary files (RC bug #639)
    - Don't explicitly link to libgcc_s
    - Add documentation for '-t' (test hook) option

1.3.1.153 [2008-10-31]
    - VAS bugs 15691, 15690, 15790: install script fixes
    - VAS bug 15586: support universal binaries for Mac OS X
    - correct conf path to /etc/opt/quest instead of /opt/quest/etc
    - improved diagnostics of config file loading
    - fix seg faults related to -s option and when Mac wakes from sleep
    - better detection of internet tools when building
    - documentation fixes
1.3.0.118 [2008-09-29]
    - alter packet format to match MS DNS Client
    - new VGP xlator for group policy control
    - use syslog for logging
    - bug 570: -t0 wasn't deleting records
    - use configuration file for options
    - follow NS records when primary SOA update fails
    - bug 484: detect the authoritative domain name by doing an SOA query
    - bug 467: add -r option for PTR updates
    - bug 415: Debian and Ubuntu support
    - fix option processing bug
1.2.1.48 [2007-07-11]
    - only support DHCP hooks in Solaris 10 and later
    - bug 189: solaris hook used wrong path to dnsupdate
    - added -N option (no-authenticate)
    - bug 72: long TSIG identifier causes refused updates; -I flag reverts
    - improve portability to Solaris 2.6, package name is now QSFTdnsup
    - improve portability to older versions of VAS 3
1.2.0.30 [2006-08-07]
    - add hook install script
    - improve documentation, portability, error messages
    - add -I option for strict RFC operation
1.0.0.12 [2006-05-10]
    - initial release


ipwatchd/ipwatchd.c
/* (c) 2008 Quest Software, Inc. All rights reserved. */

#include <stdio.h>
#include <assert.h>
#include <unistd.h>
#include <stdlib.h>
#include </usr/include/err.h>
#include <launch.h>
#include <SystemConfiguration/SystemConfiguration.h>

#include "watch.h"

/*
 * ipwatchd is a Mac OS X daemon that watches for IP address changes and
 * then executes a given command.
 * 
 * It is a replacement for the 'kicker' daemon which was removed in 
 * Mac OS 10.5 (Leopard).
 * 
 * This daemon watches the following dynamic store string
 *   State:/Network/Global/IPv4[PrimaryService]
 * to determine what the current "primary" interface (eg "0", meaning en0)
 * From this it watches the primary service's IP address(es) at
 *   State:/Network/Service/$PrimaryService/IPv4[Addresses] 
 * And uses the first IP address as the primary address of the host.
 *
 * Changes in the primary address of the host cause an external program
 * to be run.
 */

static void checkin(void);
static Boolean strings_equal(CFTypeRef s1, CFTypeRef s2);
static void notify_global_ipv4_change(void);
static void set_primary_service(CFStringRef newPrimaryService);
static Boolean substitute_pattern(const char *pattern, 
	const char *substitution, char *buf, int bufsz);
static void notify_primary_service_change(void);
static void notify_primary_ipv4_change(void);
static void set_primary_address(CFStringRef newPrimaryAddress);
static void notify_primary_address_change(void);
static void run_change_program(CFRunLoopTimerRef timer, void *info);

/* Configuration */
static const char *GlobalIPv4Key;
static const char *PrimaryIPv4KeyPattern;
static const char *ChangeProgram;
static double      ChangeDelay = -1;			  /* seconds */
static int         debug;

static SCDynamicStoreRef DynamicStore;
static CFRunLoopTimerRef ChangeTimer;

/* Cached strings from configuration */
static CFStringRef PrimaryService = NULL; /* "0" */
static CFStringRef PrimaryIPv4Key = NULL; /* "State:/Network/Service/0/IPv4" */
static CFStringRef PrimaryAddress = NULL; /* "1.2.3.4" */

static int
to_boolean(const char *s)
{
    if (strcasecmp(s, "true") == 0 || 
        strcasecmp(s, "on") == 0 || 
        strcasecmp(s, "yes") == 0)
        return TRUE;
    if (strcasecmp(s, "false") == 0 || 
        strcasecmp(s, "off") == 0 ||
        strcasecmp(s, "no"))
        return FALSE;
    warnx("unknown boolean '%.100s'", s);
    return FALSE;
}

/* Sets an option named name to the value. Returns TRUE on success */
static int
set_option(const char *name, const char *value)
{
    if (debug)
        fprintf(stderr, "set_option %s = %s\n", name, value);
    if (strcmp(name, "ChangeProgram") == 0)
        ChangeProgram = value;
    else if (strcmp(name, "GlobalIPv4Key") == 0)
        GlobalIPv4Key = value;
    else if (strcmp(name, "PrimaryIPv4KeyPattern") == 0)
        PrimaryIPv4KeyPattern = value;
    else if (strcmp(name, "ChangeDelay") == 0)
        ChangeDelay = atof(value);
    else if (strcmp(name, "Debug") == 0)
        debug = to_boolean(value);
    else
        warnx("ignoring unknown option '%.100s'", name);
    return TRUE;
}

/* Split the "name=value" string on the first '=' and set the option */
static int
set_option_from_arg(const char *arg)
{
    char opt_name[100];
    int i;

    /* Copy the beginning of arg up to the '=' into opt_name[] */
    for (i = 0; arg[i] != '='; i++) {
        if (arg[i] == 0) {
            warnx("missing '=' in -o argument '%.100s'", arg);
            return FALSE;
        }
        if (i >= sizeof opt_name - 1) {
            warnx("option name too long");
            return FALSE;
        }
        opt_name[i] = arg[i];
    }
    opt_name[i] = 0;
    return set_option(opt_name, arg + i);
}

int
main(int argc, char * const argv[])
{
	int error = 0;
	int ch;
	extern int optind;
	CFStringRef key;

	while ((ch = getopt(argc, argv, "do:")) != -1)
		switch (ch) {
		case 'd':
		    set_option("Debug", "true");
		    break;
                case 'o':
                    if (!set_option_from_arg(optarg))
                        error = 1;
                    break;
		case '?':
		    error = 1;
		    break;
		}
	if (optind < argc)
                set_option("ChangeProgram", argv[optind++]);

	if (optind < argc)
		error = 1;
	if (error) {
		fprintf(stderr, 
                        "usage: %s [-d] [-o opt=value] [ChangeProgram]\n", 
                        argv[0]);
		exit(1);
	}

	checkin();

	/* Set defaults */
	if (!GlobalIPv4Key)
		GlobalIPv4Key = "State:/Network/Global/IPv4";
	if (!PrimaryIPv4KeyPattern)
		PrimaryIPv4KeyPattern = "State:/Network/Service/%/IPv4";
	if (ChangeDelay < 0)
		ChangeDelay = 30;

	/* Display the configuration results */
	if (debug) {
		fprintf(stderr, "ChangeProgram: %s\n",
		    ChangeProgram ? ChangeProgram : "(null)");
		fprintf(stderr, "GlobalIPv4Key: %s\n", GlobalIPv4Key);
		fprintf(stderr, "PrimaryIPv4KeyPattern: %s\n", 
		    PrimaryIPv4KeyPattern);
		fprintf(stderr, "ChangeDelay: %f\n", ChangeDelay);
	}

	/* Open SC's dynamic store, which is updated by configd */
	DynamicStore = SCDynamicStoreCreate(NULL, 
	    CFSTR("com.quest.rc.ipwatchd"), watch_store_callback, NULL);
	if (!DynamicStore)
		errx(1, "SCDynamicStoreCreate: %s", SCErrorString(SCError()));

	/* Subscribe to changes in State:/Network/Global/IPv4 
	 * which indicates a change in primary interface */
	watch_init(DynamicStore);

	key = CFStringCreateWithCString(NULL, GlobalIPv4Key, 
	    kCFStringEncodingMacRoman);
	if (!key)
		errx(1, "CFStringCreateWithCString");
	watch_add_key(key, notify_global_ipv4_change);
	CFRelease(key);

	/* Connect the dynamic store notifications into the current run loop */
	CFRunLoopAddSource(CFRunLoopGetCurrent(),
	    SCDynamicStoreCreateRunLoopSource(NULL, DynamicStore, 0),
	    kCFRunLoopCommonModes);

	/* Trigger a change notification now */
	notify_global_ipv4_change();

	/* Wait for events */
	CFRunLoopRun();
	exit(0);
}

/* Check-in with launchd and receive parameters */
static void
checkin()
{
	launch_data_t checkin, response, param;

	/* Check-in with launchd */

	if (!(checkin = launch_data_new_string(LAUNCH_KEY_CHECKIN))) {
		warnx("launch_data_new_string");
		return;
	}

	if (!(response = launch_msg(checkin))) {
		warn("launch_msg");
		return;
	}
	if (launch_data_get_type(response) == LAUNCH_DATA_ERRNO) {
		warnx("launch check-in failed: %s",
			strerror(launch_data_get_errno(response)));
		return;
	}
	if (launch_data_get_type(response) != LAUNCH_DATA_DICTIONARY) {
		warnx("launch check-in failed: bad response");
		return;
	}
}

/* Returns true if two string pointers are either both NULL, or the same text.
 */
static Boolean
strings_equal(CFTypeRef s1, CFTypeRef s2)
{
	if (!s1 && !s2)
		return TRUE;
	if (!s1 || !s2)
		return FALSE;
	return CFEqual(s1, s2);
}

/*
 * Called when State:/Network/Global/IPv4 properties may have changed.
 * Checks to see if the Primary Service has changed
 */
static void
notify_global_ipv4_change()
{
	CFPropertyListRef GlobalIPv4Prop;
	CFTypeRef newPrimaryService = NULL;
	CFStringRef key;

	if (debug)
		fprintf(stderr, "notify_global_ipv4_change\n");

	key = CFStringCreateWithCString(NULL, GlobalIPv4Key, 
	    kCFStringEncodingMacRoman);
	if (!key)
		errx(1, "CFStringCreateWithCString");

	/*  get State:/Network/Global/IPv4[PrimaryService] */
	GlobalIPv4Prop = SCDynamicStoreCopyValue(DynamicStore, key);
	if (GlobalIPv4Prop)
		newPrimaryService = CFDictionaryGetValue(GlobalIPv4Prop, 
		    CFSTR("PrimaryService"));

	set_primary_service(newPrimaryService);

	CFRelease(key);
}

/* Set PrimaryService to a new string */
static void
set_primary_service(CFStringRef newPrimaryService)
{
	if (debug) {
		fprintf(stderr, "PrimaryService ");
		if (PrimaryService)
		    CFShow(PrimaryService); 
		else
		    fprintf(stderr, "(null)\n");
	}

	/* Ignore if the primary service doesn't change */
	if (strings_equal(PrimaryService, newPrimaryService))
		return;

	if (PrimaryService)
		CFRelease(PrimaryService);
	PrimaryService = newPrimaryService;
	if (PrimaryService)
		CFRetain(PrimaryService);
	notify_primary_service_change();
}

/*
 * Called when PrimaryService changes.
 * Updates subscriptions and then checks for a new address.
 */
static void
notify_primary_service_change()
{
	char service[2048];
	char key[2048];

	if (debug)
		fprintf(stderr, "notify_primary_service_change\n");

	if (PrimaryIPv4Key) {
		watch_remove_key(PrimaryIPv4Key);
		CFRelease(PrimaryIPv4Key);
		PrimaryIPv4Key = NULL;
	}

	if (!PrimaryService) {
		/* We lost the primary service, so we lose the address */
		set_primary_address(NULL);
		return;
	}

	/* Convert PrimaryService to a C string */
	if (!CFStringGetCString(PrimaryService, service, sizeof service,
	    kCFStringEncodingMacRoman))
	{
		warnx("CFStringGetCString(PrimaryService)");
		return;
	}

	/* Substitute the pattern with the primary service, to get
	 * "State:/Network/Service/${PrimaryService}/IPv4" */
	if (!substitute_pattern(PrimaryIPv4KeyPattern, service,
	    key, sizeof key))
	{
		warnx("substitute_pattern: string too long");
		return;
	}

	/* Convert substituted key into a CF string */
	PrimaryIPv4Key = CFStringCreateWithCString(NULL, key, 
	    kCFStringEncodingMacRoman);
	if (!PrimaryIPv4Key)
		errx(1, "CFStringCreateWithCString");

	/* Start watching the new Primary service's IPv4 state key */
	watch_add_key(PrimaryIPv4Key, notify_primary_ipv4_change);

	/* Trigger an immediate change notification */
	notify_primary_ipv4_change();
}

/* Substitute the first '%' in a pattern with a string.
 * Returns a buffer containing the pattern but with the first '%' replaced
 * with the content of the substitution string.
 */
static Boolean
substitute_pattern(const char *pattern, const char *substitution, 
	char *buf, int bufsz)
{
	const char *p;
	const char *s;
	int bufpos = 0;

	for (p = pattern; *p; p++)
		if (*p == '%' && substitution) {
		    for (s = substitution; *s; s++) {
			if (bufpos >= bufsz)
			    return FALSE;
			buf[bufpos++] = *s;
		    }
		    substitution = NULL;
		} else {
		    if (bufpos >= bufsz)
			return FALSE;
		    buf[bufpos++] = *p;
		}

	if (bufpos >= bufsz)
		return FALSE;

	buf[bufpos] = 0;
	return TRUE;
}

/*
 * Called when we think the State:/Network/Service/${PrimaryService}/IPv4
 * property has changed.
 */
static void
notify_primary_ipv4_change()
{
	CFPropertyListRef prop;
	CFArrayRef addresses;
	CFTypeRef newPrimaryAddress = NULL;

	if (debug)
		fprintf(stderr, "notify_primary_ipv4_change\n");

	assert(PrimaryIPv4Key != NULL);

	/*  get State:/Network/Service/${primaryService}/IPv4[Addresses] */
	prop = SCDynamicStoreCopyValue(DynamicStore, PrimaryIPv4Key);
	if (!prop)
		warnx("SCDynamicStoreCopyValue: %s", SCErrorString(SCError()));
	else {
		addresses = CFDictionaryGetValue(prop, CFSTR("Addresses"));
		if (addresses && CFArrayGetCount(addresses) > 0)
		    newPrimaryAddress = CFArrayGetValueAtIndex(addresses, 0);
	}

	set_primary_address(newPrimaryAddress);
	if (prop)
	    CFRelease(prop);
}

/*
 * Sets PrimaryAddress to the given value
 */
static void
set_primary_address(CFStringRef newPrimaryAddress)
{
	if (debug) {
		fprintf(stderr, "PrimaryAddress ");
		if (PrimaryAddress)
		    CFShow(PrimaryAddress); 
		else
		    fprintf(stderr, "(null)\n");
	}

	/* Ignore if the address doesn't change */
	if (strings_equal(PrimaryAddress, newPrimaryAddress))
		return;

	if (PrimaryAddress)
		CFRelease(PrimaryAddress);
	PrimaryAddress = newPrimaryAddress;
	if (PrimaryAddress)
		CFRetain(PrimaryAddress);
	notify_primary_address_change();
}

/*
 * Called when PrimaryAddress changes its value.
 * Arranges for the change timer to start and to call run_change_program()
 * on expiration.
 */
static void
notify_primary_address_change()
{
	CFGregorianUnits delay;

	if (debug)
		fprintf(stderr, "notify_primary_address_change\n");

	/* Start a timer to fire a short period after the address
	 * change is detected.
	 */

	if (ChangeTimer) {
		CFRunLoopTimerInvalidate(ChangeTimer);
		CFRelease(ChangeTimer);
		ChangeTimer = NULL;
	}

	/*
	 * If the primary address disappeared, then 
	 * we probably don't have a link. So forget it.
	 */
	if (!PrimaryAddress)
		return;

	/* If there is no delay required, then fire immediately */
	if (ChangeDelay <= 0) {
		run_change_program(NULL, NULL);
		return;
	}

	/* Construct a timer that will run for ChangeDelay seconds */
	memset(&delay, 0, sizeof delay);
	delay.seconds = ChangeDelay;
	ChangeTimer = CFRunLoopTimerCreate(NULL, 
	    CFAbsoluteTimeAddGregorianUnits(
		CFAbsoluteTimeGetCurrent(),
		NULL, delay),
	    0, 0, 0, run_change_program, NULL);
	if (!ChangeTimer)
		warnx("CFRunLoopTimerCreate");
	else
		CFRunLoopAddTimer(CFRunLoopGetCurrent(), 
		    ChangeTimer, kCFRunLoopCommonModes);
}

/*
 * Runs the change program, passing it the current PrimaryAddress
 * as the only argument. 
 * This function is called when the change timer expires.
 */
static void
run_change_program(CFRunLoopTimerRef timer, void *info)
{
	char address[1024];

	if (debug)
		fprintf(stderr, "run_change_program\n");

	assert(PrimaryAddress != NULL);

	if (debug) {
		fprintf(stderr, "new address: ");
		CFShow(PrimaryAddress); 
	} 

	if (!CFStringGetCString(PrimaryAddress, address,
	    sizeof address, kCFStringEncodingMacRoman))
	{
		warnx("CFStringGetCString");
		return;
	}
	
	if (ChangeProgram) {
		if (signal(SIGCHLD, SIG_IGN) == SIG_ERR)
		    warn("signal SIGCHLD");
		switch (fork()) {
		case -1: 
		    warn("fork");
		    break;
		case 0:
		    execlp(ChangeProgram, ChangeProgram, address, NULL);
		    warn("execvp: %s", ChangeProgram);
		    _exit(1);
		}
	}
}
